### ISSUE TITLE
Bad performance with mismatching sound sampling rates

### ISSUE THREAD

**Issue by quasilyte** [hajimehoshi/ebiten#2509]
### Ebitengine Version

v2.4.9

### Operating System

- [ ] Windows
- [ ] macOS
- [X] Linux
- [ ] FreeBSD
- [ ] OpenBSD
- [ ] Android
- [ ] iOS
- [ ] Nintendo Switch
- [ ] Xbox
- [ ] Web Browsers

### Go Version (`go version`)

1.18.1

### What steps will reproduce the problem?

1. Create an audio context with sampling rate of 44100
2. Find a wav sound with a sampling rate of 22050
3. Use Decode or DecodeWithSampleRate, specify 44100 as a desired sample rate
4. Create the audio player for that `*wav.Stream`
5. Play the sound, rewind, play again, measure the time

### What is the expected result?

One of the following:

* Resampling happens only once
* The resampling price is not noticeable even if it happens multiple times
* There is an API to tell whether the wav needs resampling or not, so program can react according to that

### What happens instead?

It looks like there is no resampling caching. Every rewind+play results in a resampling round.

Playing a sound without mismatching sample rate takes ~100µs on my machine. Which is good.
Playing a 22050 sample rate sound results in ... 20000µs for every play, this is ~0.02s (200 times slower than above).
The game becomes sluggish and there are visible hiccups in the game due to that.

Here is a profiling results of the game that had performance issues due to that:
![image](https://user-images.githubusercontent.com/6286655/209874749-c076972a-fa21-45c3-b2dc-c6b882f8d7d3.png)

Note that top5 is occupied by audio-related issues (resampling).

### Anything else you feel useful to add?

Potential solutions:

1. Make it possible to resample the sound only once, play+rewind should be efficient. We can use a loading screen time to resample the resources. The game scene should not experience 0.02 delays before playing a sound.

2. Add some APIs to make it easier (and possible) to detect wav resources that do not have a sample rate that matches the desired sample rate. I don't see and APIs for it right now. `*wav.Stream` doesn't provide any, `Decode` functions don't return that info either. With such a way to detect these sounds, it would be easier to resample all files on our own, before even running the game. But it's not convenient to check each file manually, especially when you're working on multiple games and sounds come from different sources. I would add a check that would print to a console if any of the sounds need to be resampled.

Ideally, I would like to have a way to re-encode the sound in the desired sample rate. So I could open the wav file, figure out that sample rate is wrong, re-encode it, then create an audio player on that re-encoded wav. Maybe there is a way to do that but I'm not aware of that option? Please let me know if that's the case.

**Comment by quasilyte** [comment]
It's not really a bug, but it's no a feature request either. I'm a little bit confused with categories, so feel free to fix it if I missed the right one.

**Comment by quasilyte** [comment]
One workaround that I found is this:

```go
stream := sys.loader.LoadWAV(id)
data, err := io.ReadAll(stream)
if err != nil {
    // handle error
}
player := sys.audioContext.NewPlayerFromBytes(data)
```

We pay the price for the resampling only once, during the ReadAll.

All in all, it looks it's exactly what I wanted to get.

It's a little bit confusing that it's as easy to end up with the wrong solution here.
Maybe we should mention something about it in the documentation?

I think this issue is useful in a sense that it'll bring people with the same issue here. They can find a solution + discussion here as well.

**Comment by hajimehoshi** [comment]
Actually Real-time resampling is not recommended now, and you should prepare the matched samplerate.

> Make it possible to resample the sound only once, play+rewind should be efficient. We can use a loading screen time to resample the resources. The game scene should not experience 0.02 delays before playing a sound.

This would mean the resampled result would be cached, and this might be too big, which is my concern.

> Add some APIs to make it easier (and possible) to detect wav resources that do not have a sample rate that matches the desired sample rate. I don't see and APIs for it right now. *wav.Stream doesn't provide any, Decode functions don't return that info either. With such a way to detect these sounds, it would be easier to resample all files on our own, before even running the game. But it's not convenient to check each file manually, especially when you're working on multiple games and sounds come from different sources. I would add a check that would print to a console if any of the sounds need to be resampled.

It's OK to have such a API, but again, I recommend preparing a wave file that has the matched samplerate.

> Ideally, I would like to have a way to re-encode the sound in the desired sample rate.

https://pkg.go.dev/github.com/hajimehoshi/ebiten/v2/audio#Resample Does this work?

**Comment by quasilyte** [comment]
> Actually Real-time resampling is not recommended now, and you should prepare the matched samplerate.

Good. This is probably the answer I was looking for.

I'll probably add some makefile scripts that would traverse the assets folder and see if any audio files have a mismatching sample rate. I think the assets management thing is out of scope for ebitengine (at least at this point?), so it's not an issue on the ebitengine part.

I was a little bit surprised to see a significant slowdown in the game and it took me some time to figure out what was causing that.

I think we can close the issue now. :)

Thank you for the quick response.
